iceoryx2-bb/cxx/tests/src/optional_tests.cpp

```diff
@@ -0,0 +1,867 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "iox2/container/optional.hpp"
+
+#include <gtest/gtest.h>
+
+namespace {
+namespace testing {
+int g_ObservableWasInitialized = 0;
+int g_ObservableWasCopied = 0;
+int g_ObservableWasCopyAssigned = 0;
+int g_ObservableWasMoved = 0;
+int g_ObservableWasMoveAssigned = 0;
+int g_ObservableWasDestructed = 0;
```

Comment:
Could we also add here an instance counter? that is incremented in every constructor and decremented in the dtor.

Reply:
I can add that for ease of understanding. The value of the instance counter would always be the sum of all the constructor counters though (`WasInitialized` + `WasMoved` + `WasCopied`), so it won't assert any properties that are not already covered.